File: SkylinkSample/SkylinkSample/MultiVideoCallViewController.m
//
//  MultiVideoCallViewController.m
//  Skylink_Examples
//
//  Created by Temasys on 11/01/2016.
//  Copyright © 2016 Temasys. All rights reserved.
//

#import "MultiVideoCallViewController.h"
#import <AVFoundation/AVFoundation.h>
#import "Constant.h"
#import "Peer.h"
//#define ROOM_NAME [[NSUserDefaults standardUserDefaults] objectForKey:@"ROOMNAME_MULTIVIDEOCALL"]
#define MY_PEER_ID @"myPeerId"

@interface MultiVideoCallViewController () <UIPickerViewDataSource, UIPickerViewDelegate, SKYLINKConnectionRecordingDelegate, SKYLINKConnectionMediaDelegate>
// IBOutlets
@property (weak, nonatomic) IBOutlet UIView *localVideoContainerView;
@property (weak, nonatomic) IBOutlet UIView *firstPeerVideoContainerView;
@property (weak, nonatomic) IBOutlet UIView *secondPeerVideoContainerView;
@property (weak, nonatomic) IBOutlet UIView *thirdPeerVideoContainerView;
@property (weak, nonatomic) IBOutlet UILabel *firstPeerLabel;
@property (weak, nonatomic) IBOutlet UILabel *secondPeerLabel;
@property (weak, nonatomic) IBOutlet UILabel *thirdPeerLabel;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *activityIndicator;
@property (weak, nonatomic) IBOutlet UIButton *lockButton;
@property (weak, nonatomic) IBOutlet UISegmentedControl *videoAspectSegmentControl;
@property (weak, nonatomic) IBOutlet UIPickerView *pickerView;
@property (weak, nonatomic) IBOutlet UIView *pickerViewContainer;
@property (weak, nonatomic) IBOutlet UIViewController *partipationsVC;
@property (weak, nonatomic) IBOutlet UIButton *restartButton;
@property (weak, nonatomic) IBOutlet UIButton *toggleCameraButton;
// Other properties
//@property (strong, nonatomic) NSMutableArray *peerIds;
//@property (strong, nonatomic) NSMutableDictionary *peersInfos;
//@property (nonatomic, copy) NSString *cameraMediaID;
//@property (nonatomic, copy) NSString *audioMediaID;
//@property (nonatomic, strong) NSMutableArray<Peer *> *peerObjects;
@property (nonatomic, strong) IBOutletCollection(UIView) NSArray *videoContainers;
@end

@implementation MultiVideoCallViewController {
    BOOL isRoomLocked;
    BOOL isRecording;
    BOOL isLocalCameraRunning;
    NSMutableArray<Peer *> *_peers;
}
#pragma mark - Init

- (void)viewDidLoad {
    [super viewDidLoad];
    MyLog(@"SKYLINKConnection version = %@", [SKYLINKConnection getSkylinkVersion]);
    self.title = @"Multi Party Video Call";
    [self setupData];
}

- (void)setupData{
    //creating SKYLINKConnectionConfig
    _roomName = ROOM_MULTI_VIDEO;
    SKYLINKConnectionConfig *config = [SKYLINKConnectionConfig new];
    [config setAudioVideoReceiveConfig:AudioVideoConfig_AUDIO_AND_VIDEO];
    [config setAudioVideoSendConfig:AudioVideoConfig_AUDIO_AND_VIDEO];
    config.isMultiTrackCreateEnable = YES;
    config.isMirrorLocalFrontCameraView = YES;
    //creating SKYLINKConnection
    _skylinkConnection = [[SKYLINKConnection alloc] initWithConfig:config callback:nil];
    _skylinkConnection.lifeCycleDelegate = self;
    _skylinkConnection.mediaDelegate = self;
    _skylinkConnection.remotePeerDelegate = self;
    _skylinkConnection.enableLogs = YES;
    //init variables
    _peers = [NSMutableArray new];
    
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self startLocalMediaDevice:SKYLINKMediaDeviceCameraFront];
    });
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [self joinRoom];
    });
}

#pragma mark - SKYLINKConnectionLifeCycleDelegate
- (void)connectionDidConnectToRoomSuccessful:(SKYLINKConnection *)connection{
    [_videoContainers.firstObject setAlpha:1.0];
    [self.activityIndicator stopAnimating];
}
- (void)connection:(SKYLINKConnection *)connection didConnectToRoomFailed:(NSString *)errorMessage{
    showAlert(@"Connection failed!", errorMessage);
}
- (void)connection:(SKYLINKConnection *)connection didDisconnectFromRoomWithSkylinkEvent:(NSDictionary *)skylinkEvent contextDescription:(NSString *)contextDescription{
    [self.activityIndicator stopAnimating];
}

#pragma mark - SKYLINKConnectionMediaDelegate
- (void)connection:(SKYLINKConnection *)connection didCreateLocalMedia:(SKYLINKMedia *)localMedia{
    NSLog(@"didCreateLocalMedia");
    //add media which is video only
    if (!localMedia) {
        return;
    }
    if (localMedia.skylinkMediaType != SKYLINKMediaTypeAudio) {
        [self addMedia:localMedia peerId:MY_PEER_ID];
        [self reloadVideoView];
    }
}
- (void)connection:(SKYLINKConnection *)connection didChangeLocalMedia:(SKYLINKMedia *)localMedia{
    NSLog(@"changed local media %d", localMedia.skylinkMediaState);
    [self.activityIndicator stopAnimating];
}
- (void)connection:(SKYLINKConnection *)connection didReceiveRemoteMedia:(SKYLINKMedia *)remoteMedia remotePeerId:(NSString *)remotePeerId{
    NSLog(@"didReceiveRemoteMedia: %u", remoteMedia.skylinkMediaType);
    if (!remoteMedia || !remotePeerId) {
        return;
    }
    //add media which is video only
    if (remoteMedia.skylinkMediaType != SKYLINKMediaTypeAudio) {
        [self addMedia:remoteMedia peerId:remotePeerId];
        [self reloadVideoView];
    }
}
- (void)connection:(SKYLINKConnection *)connection didChangeRemoteMedia:(SKYLINKMedia *)remoteMedia remotePeerId:(NSString *)remotePeerId{
    NSLog(@"didChangeRemoteMedia: %u", remoteMedia.skylinkMediaType);
}
- (void)connection:(SKYLINKConnection *)connection didChangeVideoSize:(CGSize)videoSize videoView:(UIView *)videoView peerId:(NSString *)peerId{
    NSLog(@"changed video size!");
    NSLog(@"peerId: %@", peerId);
    if (videoSize.height > 0 && videoSize.width > 0) {
        for (UIView* container in self.videoContainers) {
            if ([videoView isDescendantOfView:container]) {
                NSInteger _index = [self.videoContainers indexOfObject:container];
                if (_index==0) {
                    break;
                }
                _peers[_index].videoSize = videoSize;
                [videoView aspectFitRectForSize:videoSize container:container];
            }
        }
    }
}
#pragma mark - SKYLINKConnectionRemotePeerDelegate
- (void)connection:(SKYLINKConnection *)connection didReceiveRemotePeerInRoomWithRemotePeerId:(NSString *)remotePeerId userInfo:(id)userInfo{
    NSLog(@"didReceiveRemotePeerInRoomWithRemotePeerId: %@ \nuserInfo: %@", remotePeerId, userInfo);
    showAlertAutoDismiss(nil, [NSString stringWithFormat:@"%@ has joined room\n\nUserData:%@", remotePeerId, userInfo[@"userData"]], 2, self);
    if (remotePeerId){
        [self addMedia:nil peerId:remotePeerId];
        [self.pickerView reloadAllComponents];
    }
}
- (void)connection:(SKYLINKConnection *)connection didConnectWithRemotePeer:(NSString *)remotePeerId userInfo:(id)userInfo hasDataChannel:(BOOL)hasDataChannel{
//    if (remotePeerId) {
//        showAlertAutoDismiss(nil, [NSString stringWithFormat:@"%@ has joined room\n\nUserData:%@", remotePeerId, userInfo[@"userData"]], 2, self);
//        [self addMedia:nil peerId:remotePeerId];
////        [_peers addObject:[[Peer alloc] initWithPeerID:remotePeerId]];
//        [self reloadVideoView];
//        [self.pickerView reloadAllComponents];
//    }
//    [self.activityIndicator stopAnimating];
}
- (void)connection:(SKYLINKConnection *)connection didDisconnectWithRemotePeer:(NSString *)remotePeerId userInfo:(id)userInfo hasDataChannel:(BOOL)hasDataChannel{
    if (remotePeerId) {
        [self removePeer:remotePeerId];
        [self reloadVideoView];
    }
    [self.activityIndicator stopAnimating];
}
- (void)connection:(SKYLINKConnection *)connection didReceiveRemotePeerLeaveRoom:(NSString *)remotePeerId userInfo:(id)userInfo skylinkInfo:(NSDictionary *)skylinkInfo{
    
}
//- (void)connection:(SKYLINKConnection *)connection didChangeRemoteMedia:(SKYLINKMedia *)remoteMedia remotePeerId:(NSString *)remotePeerId{
//    if (remoteMedia.skylinkMediaType == SKYLINKMediaTypeVideoCamera) {
//        if (remoteMedia.skylinkMediaState == SKYLINKMediaStateUnavailable) {
//            <#statements#>
//        }
//    }
//}

#pragma mark - UIPickerView Delegate
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return _peers.count;
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    if (row == 0) return @"All";
    return [NSString stringWithFormat:@"%@:%@", _peers[row].userName, _peers[row].peerId];
}
#pragma mark - Private functions
- (void)removePeer:(NSString*)peerId{
    for (Peer *_peer in _peers) {
        if ([_peer.peerId isEqualToString:peerId]) {
            [_peers removeObject:_peer];
            return;
        }
    }
}
- (void)addMedia:(SKYLINKMedia *)media peerId:(NSString *)peerId{
    if (!peerId) {
        return;
    }
    Peer *_peer = nil;
    for (Peer *peer in _peers) {
        if ([peer.peerId isEqualToString:peerId]) {
            _peer = peer;
            break;
        }
    }
    if (!_peer) {
        _peer = [[Peer alloc] initWithPeerID:peerId];
        if ([peerId isEqualToString:MY_PEER_ID]) {
            [_peers insertObject:_peer atIndex:0];
        }else{
            [_peers addObject:_peer];
        }
    }
    if (media) {
        [_peer.medias addObject:media];
    }
//    NSLog(@"ccc===peers count: %lu", (unsigned long)_peers.count);
//    for (Peer *_peer in _peers) {
//        NSLog(@"peer: %@", _peer.peerId);
//        NSLog(@"medias count: %lu", (unsigned long)_peer.medias.count);
//        NSLog(@"first media: %d", _peer.medias.firstObject.skylinkMediaType);
//    }
}
- (void)reloadVideoView{
    //remove all videos
    for (UIView *view in _videoContainers) {
        [view removeSubviews];
    }
    //add videos from peers array
    NSLog(@"reload video. Peers count: %lu", (unsigned long)_peers.count);
    for (Peer *peer in _peers) {
        if ([peer.peerId isEqualToString:MY_PEER_ID] && peer.medias.firstObject) {
            [self addRenderedVideo:peer.medias.firstObject.skylinkVideoView insideContainer:_videoContainers.firstObject mirror:YES];
        }else{
            if (peer.medias.firstObject.skylinkMediaState!=SKYLINKMediaStateUnavailable) {
                NSInteger _index = MIN([_peers indexOfObject:peer], 3);
                NSLog(@"addview index: %ld", (long)_index);
                [self addRenderedVideo:peer.medias.firstObject.skylinkVideoView insideContainer:_videoContainers[_index] mirror:NO];
            }
        }
    }
}
- (void)addRenderedVideo:(UIView *)videoView insideContainer:(UIView *)containerView mirror:(BOOL)shouldMirror {
    [videoView aspectFitRectForSize:videoView.frame.size container:containerView];
//    [containerView addSubview:videoView];
    [containerView insertSubview:videoView atIndex:0];
}
- (void)startRecording {
    if (!_skylinkConnection.isRecording) {
        [_skylinkConnection startRecording:^(NSError * _Nullable error) {
            if (error) MyLog(@"%@", error.localizedDescription);
            showAlertAutoDismiss(nil, @"You recording is started", 3, self);
        }];
    }
}

- (void)stopRecording {
    if (_skylinkConnection.isRecording) {
        [_skylinkConnection stopRecording:^(NSError * _Nullable error) {
            if (error) MyLog(@"%@", error.localizedDescription);
            showAlertAutoDismiss(nil, @"You recording is stopped", 3, self);
        }];
    }
}

#pragma mark - IBActions
- (IBAction)toogleVideoTap:(UIButton *)sender {
    [_skylinkConnection muteVideo:!_skylinkConnection.isVideoMuted];
    [sender setImage:[UIImage imageNamed:( (_skylinkConnection.isVideoMuted) ? @"NoVideoFilled.png" : @"VideoCall.png")] forState:UIControlStateNormal];
}
- (IBAction)toogleSoundTap:(UIButton *)sender {
    [_skylinkConnection muteAudio:!_skylinkConnection.isAudioMuted];
    [sender setImage:[UIImage imageNamed:( (_skylinkConnection.isAudioMuted) ? @"NoMicrophoneFilled.png" : @"Microphone.png")] forState:UIControlStateNormal];
}
- (IBAction)switchCameraTap:(UIButton *)sender {
    [_skylinkConnection switchCamera:nil];
}
- (IBAction)lockRoom:(UIButton *)sender {
    isRoomLocked ? [_skylinkConnection unlockTheRoom:nil] : [_skylinkConnection lockTheRoom:nil];
    isRoomLocked = !isRoomLocked;
    [self.lockButton setImage:[UIImage imageNamed:isRoomLocked ? @"LockFilled" : @"Unlock.png"] forState:UIControlStateNormal];
}

- (IBAction)videoAspectSegmentControlChanged:(UISegmentedControl *)sender {
//    [self updatePeersVideosFrames];
}

- (IBAction)recording:(UISwitch *)sender {
    sender.isOn ? [self startRecording] : [self stopRecording];
}

- (IBAction)restart {
    [self.pickerViewContainer setHidden:NO];
}

- (IBAction)toolbarDone {
    [self.pickerViewContainer setHidden:YES];
}

- (IBAction)toolbarSend {
    [self.pickerViewContainer setHidden:YES];
    if ([self.pickerView selectedRowInComponent:0] == 0) {
        [_skylinkConnection refreshConnectionWithRemotePeerId:nil doIceRestart:YES callback:nil];
    }else{
        NSString *_selectedPeerId = _peers[[_pickerView selectedRowInComponent:0]].peerId;
        [_skylinkConnection refreshConnectionWithRemotePeerId:_selectedPeerId doIceRestart:YES callback:nil];
    }
}
@end

File: Swift_version/sampleapp/Sample_App/Sample_App/MultiVideoCallViewController.swift
//
//  MultiVideoCallViewController.swift
//  Sample_App
//
//  Created by Phyo Pwint  on 7/4/16.
//  Copyright © 2016 Temasys . All rights reserved.
//

import UIKit
import AVFoundation


class MultiVideoCallViewController: UIViewController, SKYLINKConnectionLifeCycleDelegate, SKYLINKConnectionMediaDelegate, SKYLINKConnectionRemotePeerDelegate {
    
    var skylinkConnection: SKYLINKConnection!
    var peerIds = [String]()
    var peersInfos : NSMutableDictionary!
    var skylinkApiKey: String!
    var skylinkApiSecret: NSString!
    
    @IBOutlet weak var localVideoContainerView: UIView!
    @IBOutlet weak var firstPeerVideoContainerView: UIView!
    @IBOutlet weak var secondPeerVideoContainerView: UIView!
    @IBOutlet weak var thirdPeerVideoContainerView: UIView!
    @IBOutlet weak var firstPeerLabel: UILabel!
    @IBOutlet weak var secondPeerLabel: UILabel!
    @IBOutlet weak var thirdPeerLabel: UILabel!
    @IBOutlet weak var lockButton: UIButton!
    @IBOutlet weak var btnAudioTap: UIButton!
    @IBOutlet weak var btnFlipCamera: UIButton!
    @IBOutlet weak var btnCameraTap: UIButton!
    @IBOutlet weak var videoAspectSegmentControl: UISegmentedControl!
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
    
    let ROOM_NAME = "MULTI-VIDEO-CALL-ROOM"
    var isRoomLocked: Bool!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        NSLog("imat_viewDidLoad")
        NSLog("SKYLINKConnection version = %@", SKYLINKConnection.getSkylinkVersion())
        self.title = "Multi Party Video Call"
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(image: UIImage(named: "Cancel.png"), style: .Plain, target: self, action: #selector(MultiVideoCallViewController.disConnect))
        let infoButton: UIButton = UIButton(type: .InfoLight)
        infoButton.addTarget(self, action: #selector(MultiVideoCallViewController.showInfo), forControlEvents: .TouchUpInside)
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: infoButton)
        
        self.peersInfos = [:]
        self.isRoomLocked = false
        
        // Creating configuration
        let config:SKYLINKConnectionConfig = SKYLINKConnectionConfig()
        config.video = true
        config.audio = true
        
        // Creating SKYLINKConnection
        self.skylinkConnection = SKYLINKConnection(config: config, appKey: self.skylinkApiKey! as String)
        self.skylinkConnection.lifeCycleDelegate = self
        self.skylinkConnection.mediaDelegate = self
        self.skylinkConnection.remotePeerDelegate = self
        
        // Connecting to a room
        SKYLINKConnection.setVerbose(true)
        
        self.skylinkConnection.connectToRoomWithSecret(self.skylinkApiSecret! as String, roomName: ROOM_NAME, userInfo: nil)
        
        //Disable Button
        btnFlipCamera.enabled = false
        btnAudioTap.enabled = false
        btnCameraTap.enabled = false
        lockButton.enabled = false
        
    }
    
    func disConnect() {
        NSLog("imat_disConnect")
        self.activityIndicator.startAnimating()
        if (self.skylinkConnection != nil) {
            self.skylinkConnection.disconnect({() -> Void in
                self.activityIndicator.stopAnimating()
                self.navigationController!.popViewControllerAnimated(true)
            })
        }
    }
    
    func showInfo() {
        let msgTitle: String = "Infos"
        let msg: String = "\nRoom name:\n\(ROOM_NAME)\n\nLocal ID:\n\(self.skylinkConnection.myPeerId)\n\nKey: •••••\(self.skylinkApiKey.substringFromIndex(self.skylinkApiKey.startIndex.advancedBy(self.skylinkApiKey.characters.count-7)))\n\nSkylink version \(SKYLINKConnection.getSkylinkVersion())"
        AlertMessage(msgTitle, msg:msg)
    }
    
    func AlertMessage(msg_title: String, msg:String) {
        let alertController = UIAlertController(title: msg_title , message: msg, preferredStyle: .Alert)
        let OKAction = UIAlertAction(title: "OK", style: .Default) { (action) in }
        alertController.addAction(OKAction)
        self.presentViewController(alertController, animated: true) {
        }
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        self.updatePeersVideosFrames()
    }
    
    // SKYLINK SDK Delegates methods
    
    // MARK: SKYLINKConnectionMediaDelegate
    
    func connection(connection: SKYLINKConnection, didChangeVideoSize videoSize: CGSize, videoView: UIView) {
        
        if videoSize.height > 0 && videoSize.width > 0 {
            let correspondingContainerView: UIView = self.containerViewForVideoView(videoView)
            if !correspondingContainerView.isEqual(self.localVideoContainerView) {
                let i: Int = self.indexForContainerView(correspondingContainerView)
                
                
                let videoView : AnyObject = (self.peersInfos[peerIds[i]]?["videoView"])!
                let videoSize = NSValue(CGSize: videoSize)
                let isAudioMuted : AnyObject = (self.peersInfos[peerIds[i]]?["isAudioMuted"])!
                let isVideoMuted : AnyObject = (self.peersInfos[peerIds[i]]?["isVideoMuted"])!
                
                if i != NSNotFound {
                    self.peersInfos[peerIds[i]] = [
                        "videoView" : videoView,
                        "videoSize" : videoSize,
                        "isAudioMuted" : isAudioMuted,
                        "isVideoMuted" : isVideoMuted]
                }
            }
            
            videoView.frame =
                (self.videoAspectSegmentControl.selectedSegmentIndex == 0 || correspondingContainerView.isEqual(self.localVideoContainerView)) ? self.aspectFillRectForSize(videoSize, containedInRect: correspondingContainerView.frame): AVMakeRectWithAspectRatioInsideRect(videoSize, correspondingContainerView.bounds)
        }
        self.updatePeersVideosFrames()
        
    }
    
    func connection(connection: SKYLINKConnection, didToggleAudio isMuted: Bool, peerId: String) {
        
        let bool = self.peersInfos.allKeys.contains({(peerId) -> Bool in
            return true
        })
        if (bool) {
            let videoView : AnyObject = (self.peersInfos[peerId]?["videoView"])!
            let videoSize : AnyObject = (self.peersInfos[peerId]?["videoSize"])!
            let isAudioMuted = Int(isMuted)
            let isVideoMuted : AnyObject = (self.peersInfos[peerId]?["isVideoMuted"])!
            self.peersInfos[peerId] = [
                "videoView" : videoView,
                "videoSize" : videoSize,
                "isAudioMuted" : isAudioMuted,
                "isVideoMuted" : isVideoMuted]
        }
        self.refreshPeerViews()
        
    }
    
    func connection(connection: SKYLINKConnection, didToggleVideo isMuted: Bool, peerId: String) {
        NSLog("imat_didToggleVideo")
        let bool = self.peersInfos.allKeys.contains({(peerId) -> Bool in
            return true
        })
        
        if (bool) {
            let videoView : AnyObject = (self.peersInfos[peerId]?["videoView"])!
            let videoSize : AnyObject = (self.peersInfos[peerId]?["videoSize"])!
            let isAudioMuted : AnyObject = (self.peersInfos[peerId]?["isAudioMuted"])!
            let isVideoMuted = Int(isMuted)
            self.peersInfos[peerId] = [
                "videoView" : videoView,
                "videoSize" : videoSize,
                "isAudioMuted" : isAudioMuted,
                "isVideoMuted" : isVideoMuted]
        }
        self.refreshPeerViews()
    }
    
    // MARK: SKYLINKConnectionLifeCycleDelegate
    
    func connection(connection: SKYLINKConnection, didConnectWithMessage errorMessage: String, success isSuccess: Bool) {
        
        if isSuccess {
            NSLog("Inside %@", #function)
            dispatch_async(dispatch_get_main_queue(), {() -> Void in
                self.localVideoContainerView.alpha = 1
            })
        }
        else {
            let msgTitle: String = "Connection failed"
            let msg: String = errorMessage
            AlertMessage(msgTitle, msg:msg)
            self.navigationController!.popViewControllerAnimated(true)
        }
        dispatch_async(dispatch_get_main_queue(), {() -> Void in
            self.activityIndicator.stopAnimating()
            //Enabled Btn
            self.btnFlipCamera.enabled = true
            self.btnAudioTap.enabled = true
            self.btnCameraTap.enabled = true
            self.lockButton.enabled = true
        })
    }
    
    func connection(connection: SKYLINKConnection, didLockTheRoom lockStatus: Bool, peerId: String) {
        isRoomLocked = lockStatus
        
        self.lockButton.setImage(UIImage(named: ((self.isRoomLocked.boolValue) ? "LockFilled" : "Unlock.png")), forState: .Normal)
    }
    
    
    func connection(connection: SKYLINKConnection, didRenderUserVideo userVideoView: UIView) {
        
        self.addRenderedVideo(userVideoView, insideContainer: self.localVideoContainerView, mirror: true)
    }
    
    func connection(connection: SKYLINKConnection, didDisconnectWithMessage errorMessage: String) {
        
        let msgTitle: String = "Disconnected"
        let alertController = UIAlertController(title: msgTitle , message: errorMessage, preferredStyle: .Alert)
        let OKAction = UIAlertAction(title: "OK", style: .Default) { (action) in }
        alertController.addAction(OKAction)
        self.presentViewController(alertController, animated: true, completion: {
            self.disConnect()
        })
    }
    
    // MARK: SKYLINKConnectionRemotePeerDelegate
    func connection(connection: SKYLINKConnection, didJoinPeer userInfo: AnyObject, mediaProperties pmProperties: SKYLINKPeerMediaProperties, peerId: String) {
        
        if !(self.peerIds.contains(peerId)) {
            self.peerIds.append(peerId)
        }
        if self.peerIds.count >= 4 {
            self.lockRoom(true)
        }
        
        var bool = false
        for i in self.peersInfos.allKeys {
            if i.isEqualToString(peerId) {
                bool = true
            }
        }
        
        if !(bool) {
            self.peersInfos.addEntriesFromDictionary([peerId: ["videoView": NSNull(), "videoSize": NSNull(), "isAudioMuted": NSNull(), "isVideoMuted": NSNull()]])
        }
        
        let videoView: AnyObject = (self.peersInfos[peerId]?["videoView"])!
        let size = CGSize(width: pmProperties.videoWidth, height: pmProperties.videoHeight)
        let videoSize = NSValue(CGSize: size)
        let isAudioMuted = NSNumber(bool: pmProperties.isAudioMuted)
        let isVideoMuted = NSNumber(bool: pmProperties.isVideoMuted)
        self.peersInfos[peerId] = [
            "videoView" : videoView,
            "videoSize" : videoSize,
            "isAudioMuted" : isAudioMuted,
            "isVideoMuted" : isVideoMuted]
        self.refreshPeerViews()
        
    }
    
    func connection(connection: SKYLINKConnection, didLeavePeerWithMessage errorMessage: String, peerId: String) {
        
        NSLog("Peer with id %@ left the room with message: %@", peerId, errorMessage);
        if(self.peerIds.count != 0) {
            self.peerIds.removeAtIndex(peerIds.indexOf(peerId)!)
            self.peersInfos.removeObjectForKey(peerId)
        }
        self.lockRoom(false)
        self.refreshPeerViews()
        
    }
    
    func connection(connection: SKYLINKConnection, didRenderPeerVideo peerVideoView: UIView, peerId: String) {
        
        if !self.peerIds.contains(peerId) {
            self.peerIds.append(peerId)
        }
        var bool = false
        for i in self.peersInfos.allKeys {
            if i.isEqualToString(peerId) {
                bool = true
            }
        }
        
        if !(bool) {
            self.peersInfos.addEntriesFromDictionary([peerId: ["videoView": NSNull(), "videoSize": NSNull(), "isAudioMuted": NSNull(), "isVideoMuted": NSNull()]])
        }
        
        let videoSize: AnyObject = (self.peersInfos[peerId]?["videoSize"])!
        let isAudioMuted: AnyObject = (self.peersInfos[peerId]?["isAudioMuted"])!
        let isVideoMuted: AnyObject = (self.peersInfos[peerId]?["isVideoMuted"])!
        
        self.peersInfos[peerId] = [
            "videoView" : peerVideoView,
            "videoSize" : videoSize,
            "isAudioMuted" : isAudioMuted,
            "isVideoMuted" : isVideoMuted]
        self.refreshPeerViews()
    }
    
    //End of Skylink SDK functions
    
    // MARK: Utils
    
    func updatePeersVideosFrames() {
        var pvView: AnyObject?
        var pvSize: AnyObject?
        
        for (var i = 0; i < self.peerIds.count && i < 3; i += 1) {
            pvView = (self.peersInfos[self.peerIds[i]]?["videoView"])!
            pvSize = (self.peersInfos[self.peerIds[i]]?["videoSize"])!
            
            if (pvView is UIView && pvSize is NSValue) {
                ((pvView as! UIView)).frame = (self.videoAspectSegmentControl.selectedSegmentIndex == 0) ? self.aspectFillRectForSize((pvSize as! NSValue).CGSizeValue(), containedInRect:(self.containerViewForVideoView(pvView as! UIView).frame)): AVMakeRectWithAspectRatioInsideRect(((pvSize as! NSValue)).CGSizeValue(), self.containerViewForVideoView(pvView as! UIView).bounds)
            }
        }
    }
    
    func lockRoom(shouldLock: Bool) {
        (shouldLock) ? self.skylinkConnection.lockTheRoom() : self.skylinkConnection.unlockTheRoom()
        isRoomLocked = shouldLock
        self.lockButton.setImage(UIImage(named: ((self.isRoomLocked.boolValue) ? "LockFilled" : "Unlock.png")), forState: .Normal)
    }
    
    func containerViewForVideoView(videoView: UIView) -> UIView {
        
        var correspondingContainerView: UIView!
        
        if videoView.isDescendantOfView(self.localVideoContainerView) {
            correspondingContainerView = self.localVideoContainerView
        }
        else if videoView.isDescendantOfView(self.firstPeerVideoContainerView) {
            correspondingContainerView = self.firstPeerVideoContainerView
        }
        else if videoView.isDescendantOfView(self.secondPeerVideoContainerView) {
            correspondingContainerView = self.secondPeerVideoContainerView
        }
        else if videoView.isDescendantOfView(self.thirdPeerVideoContainerView) {
            correspondingContainerView = self.thirdPeerVideoContainerView
        }
        return correspondingContainerView
    }
    
    func aspectFillRectForSize(insideSize: CGSize, containedInRect containerRect: CGRect) -> CGRect {
        let maxFloat: CGFloat = max(containerRect.size.height, containerRect.size.width)
        let aspectRatio: CGFloat = insideSize.width / insideSize.height
        var frame: CGRect = CGRectMake(0, 0, containerRect.size.width, containerRect.size.height)
        if insideSize.width < insideSize.height {
            frame.size.width = maxFloat
            frame.size.height = frame.size.width / aspectRatio
        }
        else {
            frame.size.height = maxFloat;
            frame.size.width = frame.size.height * aspectRatio;
        }
        frame.origin.x = (containerRect.size.width - frame.size.width) / 2;
        frame.origin.y = (containerRect.size.height - frame.size.height) / 2;
        return frame
    }
    
    func addRenderedVideo(videoView: UIView, insideContainer containerView: UIView, mirror shouldMirror: Bool) {
        NSLog("I_addRenderedVideo")
        videoView.frame = containerView.bounds
        if shouldMirror {
            videoView.transform = CGAffineTransformMakeScale(-1.0, 1.0)
        }
        
        for subview: UIView in containerView.subviews {
            subview.removeFromSuperview()
        }
        
        containerView.insertSubview(videoView, atIndex: 0)
    }
    
    func indexForContainerView(v: UIView) -> Int {
        return [self.firstPeerVideoContainerView, self.secondPeerVideoContainerView, self.thirdPeerVideoContainerView].indexOfObject(v)
    }
    
    func refreshPeerViews() {
        
        let peerContainerViews = [self.firstPeerVideoContainerView, self.secondPeerVideoContainerView, self.thirdPeerVideoContainerView]
        var peerLabels = [self.firstPeerLabel, self.secondPeerLabel, self.thirdPeerLabel]
        
        
        for viewToClean: UIView in peerContainerViews {
            for aSubview: UIView in viewToClean.subviews {
                aSubview.removeFromSuperview()
            }
        }
        
        for i in 0 ..< peersInfos.count  {
            let index: Int = self.peerIds.indexOf(peerIds[i])!
            
            let videoView: AnyObject = (self.peersInfos[peerIds[i]]?["videoView"])!
            if (index < peerContainerViews.count) {
                if videoView is NSNull {
                    self.AlertMessage("Warning",msg: "Cannot render the video view. Camera not found")
                }
                else {
                    self.addRenderedVideo(videoView as! UIView, insideContainer: peerContainerViews[index], mirror: false)
                }
            }
            // refresh the label
            let audioMuted: AnyObject = (self.peersInfos[peerIds[i]]?["isAudioMuted"])!
            
            let videoMuted: AnyObject = (self.peersInfos[peerIds[i]]?["isVideoMuted"])!
            
            var mutedInfos: String = ""
            if (audioMuted is NSNumber) && CBool(audioMuted as! NSNumber) {
                mutedInfos = "Audio muted"
            }
            if (videoMuted is NSNumber) && CBool(videoMuted as! NSNumber) {
                mutedInfos = mutedInfos.characters.count != 0 ? "Video & ".stringByAppendingString(mutedInfos) : "Video muted"
            }
            if (index < peerLabels.count) {
                peerLabels[index].text = mutedInfos
            }
            if (index < peerLabels.count) {
                peerLabels[index].hidden = !(mutedInfos.characters.count != 0)
            }
        }
        
        for i in self.peerIds.count ..< peerLabels.count {
            ((peerLabels[i] as UILabel)).hidden = true
        }
        self.updatePeersVideosFrames()
    }
    
    // MARK: IB Action Function
    @IBAction func toogleVideoTap(sender: AnyObject) {
        self.skylinkConnection.muteVideo(!self.skylinkConnection.isVideoMuted())
        sender.setImage(UIImage(named: ((self.skylinkConnection.isVideoMuted()) ? "NoVideoFilled.png" : "VideoCall.png")), forState: .Normal)
        self.localVideoContainerView.hidden = (self.skylinkConnection.isVideoMuted)()
    }
    
    @IBAction func toogleSoundTap(sender: AnyObject) {
        self.skylinkConnection.muteAudio(!self.skylinkConnection.isAudioMuted())
        sender.setImage(UIImage(named: ((self.skylinkConnection.isAudioMuted()) ? "NoMicrophoneFilled.png" : "Microphone.png")), forState: .Normal)
    }
    
    @IBAction func switchCameraTap(sender: AnyObject) {
        self.skylinkConnection.switchCamera()
    }
    
    @IBAction func switchLockTap(sender: AnyObject) {
        self.lockRoom(!isRoomLocked)
    }
    
    @IBAction func videoAspectSegmentControlChanged(sender: AnyObject) {
        self.updatePeersVideosFrames()
    }
    
    
}
